### PR TITLE
workflow: codesign macOS builds

### PR DESCRIPTION
**Pull request by hjpotter92**
**What does this pull request do? Explain your changes. (required)**

The macOS binaries are signed using apple developer certificates

**Specific updates (required)**

- adds a `GO_BUILD_DIR` variable, to output built binaries to custom path
- codesign and notarize binaries when building on macos runner

**How did you test each of these updates (required)**

gh-action runs

**Checklist:**

- [x] Read the [contribution guide](./doc/contributing.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [x] README and other documentation updated
- [x] [Pending changelog](./CHANGELOG_PENDING.md) updated

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -104,25 +104,27 @@ jobs:
     runs-on: macos-latest
     steps:
       - name: Set build environment
-        run: echo "GOARCH=${{ matrix.arch }}" >> $GITHUB_ENV
+        run: |
+          echo "GOARCH=${{ matrix.arch }}" >> $GITHUB_ENV
+          echo "GO_BUILD_DIR=lp-builds/" >> $GITHUB_ENV
```
Comment (hjpotter92) [inline .github/workflows/build.yaml:111]: could not find a way to generate a random build directory name and still use it later for codesigning stage. hardcoded as `lp-builds` here until then.

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -156,6 +158,24 @@ jobs:
...
           GHA_REF: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.ref || github.ref }}
 
+      - uses: actions-ecosystem/action-regex-match@v2
+        id: match-tag
+        with:
+          text: ${{ github.event.workflow_run.head_branch }}
+          regex: '^v([0-9]+\.\d+\.\d+)$'
+
+      - name: Codesign and notarize binaries
+        if: ${{ steps.match-tag.outputs.match != '' }}
```
Comment (leszko) [inline .github/workflows/build.yaml:170]: Is this tag matching only to make sure that this is not executed for `master` builds and pull requests? If yes, then we don't we want to execute this step for `master` builds and pull requests?

**Review by leszko — COMMENTED** [review]
Added one comment. Other than that, it looks good. Thanks for the PR @hjpotter92

File: .github/workflows/build.yaml (hunk 2 above)
Comment (hjpotter92) [reply .github/workflows/build.yaml:170]: oh, completely missed that. will update the pattern to work with master branch as well run the workflow for tag push.

thanks for pointing this out!

Comment (leszko) [reply .github/workflows/build.yaml:170]: If we want to run it also on `master`, then do we need the `actions-ecosystem/action-regex-match` step at all?

Comment (hjpotter92) [reply .github/workflows/build.yaml:170]: we don't want it to sign binaries which are generated for all the commits made on an open PR. The jobs would currently run for any open PR as well as newly pushed tags and any push to the master branch:

```yaml
on:
  pull_request:
  push:
    branches:
      - master
    tags:
      - "v*"
```
The regex-match action limits codesigning to only work with `main|master` branch or a semver tag

**Review by leszko — APPROVED** [review]

**Pull request merged**